pxf/pxf-service/src/test/java/org/apache/hawq/pxf/service/UGICacheTest.java

```diff
@@ -0,0 +1,197 @@
+
+    @Test
+    public void getTwoUGIsWithDifferentSessionsForSameUser() throws Exception {
+        SessionId otherSession = new SessionId(0, "txn-id-2", "the-user");
+        UGICacheEntry proxyUGI1 = cache.getTimedProxyUGI(session);
+        UGICacheEntry proxyUGI2 = cache.getTimedProxyUGI(otherSession);
+        assertNotEquals(proxyUGI1, proxyUGI2);
+        verify(provider, times(2)).createProxyUGI("the-user");
+        // TODO: this seems weird. We're creating two UGIs with the same params,
+        // even though we have two different sessions. Why?
+    }
+
+    @Test
+    public void getTwoUGIsWithDifferentUsers() throws Exception {
+        SessionId otherSession = new SessionId(0, "txn-id", "different-user");
+        UGICacheEntry proxyUGI1 = cache.getTimedProxyUGI(session);
+        UGICacheEntry proxyUGI2 = cache.getTimedProxyUGI(otherSession);
+        assertNotEquals(proxyUGI1, proxyUGI2);
+        verify(provider, times(1)).createProxyUGI("the-user");
+        verify(provider, times(1)).createProxyUGI("different-user");
+    }
+
+    @Test
+    public void getTwoUGIsWithDifferentUsersCachesBoth() throws Exception {
+        SessionId otherSession = new SessionId(0, "txn-id", "different-user");
+        UGICacheEntry proxyUGI1a = cache.getTimedProxyUGI(session);
+        UGICacheEntry proxyUGI1b = cache.getTimedProxyUGI(session);
+        UGICacheEntry proxyUGI2a = cache.getTimedProxyUGI(otherSession);
+        UGICacheEntry proxyUGI2b = cache.getTimedProxyUGI(otherSession);
+        assertEquals(proxyUGI1a, proxyUGI1b);
+        assertEquals(proxyUGI2a, proxyUGI2b);
+        assertNotEquals(proxyUGI1a, proxyUGI2a);
+        verify(provider, times(1)).createProxyUGI("the-user");
+        verify(provider, times(1)).createProxyUGI("different-user");
+    }
+
+    @Test
+    public void getUGIWhenRequestedUserDoesNotExist() throws Exception {
+        // what does UserGroupInformation.createProxyUser() do in this scenario?
+        // how about getLoginUser()?
+    }
+
+    @Test
+    public void anySegmentIdIsValid() throws Exception {
+        session = new SessionId(65, "txn-id", "the-user");
+        UGICacheEntry proxyUGI1 = cache.getTimedProxyUGI(session);
+        assertNotNull(proxyUGI1.getUGI());
+    }
+
+    @Test
+    public void releaseWithoutForceClean() throws Exception {
+        UGICacheEntry proxyUGI1 = cache.getTimedProxyUGI(session);
+
+        cache.release(proxyUGI1, false);
+        // UGI wasn't cleaned up, so we can still get it
+        UGICacheEntry proxyUGI2 = cache.getTimedProxyUGI(session);
+        assertEquals(proxyUGI1, proxyUGI2);
+        verify(provider, times(1)).createProxyUGI("the-user");
+    }
+
+    @Test
+    public void releaseWithForceClean() throws Exception {
+        UGICacheEntry proxyUGI1 = cache.getTimedProxyUGI(session);
+
+        cache.release(proxyUGI1, true);
+        UGICacheEntry proxyUGI2 = cache.getTimedProxyUGI(session);
+        assertNotEquals(proxyUGI1, proxyUGI2);
```

Comment:
hmm.. they should be equal but not same, does actual UGI ref gets into equals() method on CacheEntry ?